### PR TITLE
ABI encode assetData fields

### PR DESCRIPTION
**Pull request by Recmo**
## Types of changes

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [ ] New feature (non-breaking change which adds functionality)
*   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [ ] Change requires a change to the documentation.
*   [ ] Added tests to cover my changes.
*   [ ] Added new entries to the relevant CHANGELOG.jsons.
*   [ ] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [ ] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/contracts/src/contracts/current/protocol/AssetProxy/interfaces/IAssetData.sol [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.4.23;
+
+interface IAssetData {
```
Comment (abandeali1) [inline packages/contracts/src/contracts/current/protocol/AssetProxy/interfaces/IAssetData.sol]: Where is this being used?

File: packages/contracts/test/asset_proxy/decoder.ts [hunk 2]
```diff
@@ -44,7 +44,7 @@ describe('TestAssetDataDecoders', () => {
         it('should correctly decode ERC721 asset data', async () => {
             const tokenId = generatePseudoRandomSalt();
             const encodedAssetData = assetProxyUtils.encodeERC721AssetData(testAddress, tokenId);
-            const encodedAssetDataWithoutProxyId = encodedAssetData.slice(0, -2);
+            console.log(encodedAssetData);
```
Comment (abandeali1) [inline packages/contracts/test/asset_proxy/decoder.ts]: Remove

File: packages/contracts/src/contracts/current/protocol/AssetProxy/interfaces/IAssetData.sol (hunk 1 above)
Comment (Recmo) [reply packages/contracts/src/contracts/current/protocol/AssetProxy/interfaces/IAssetData.sol]: I wanted to use it as 
```
bytes4 constant PROXY_ID = IAssetData.ERC20Token.selector
```
and 
```
const tokenAsset = IAssetData.ERC20Token.getABIEncodedTransactionData(tokenAddress);
```

Unfortunately neither work.

File: packages/contracts/src/contracts/current/protocol/AssetProxy/MixinERC20Transfer.sol [hunk 3]
```diff
@@ -41,36 +42,48 @@ contract MixinERC20Transfer is
...
+                gas,
+                token,
+                0,
+                cdStart,
+                sub(cdEnd, cdStart),
+                cdStart,
+                32
+            )
+            if success {
+                if returndatasize {
```
Comment (fabioberger) [inline packages/contracts/src/contracts/current/protocol/AssetProxy/MixinERC20Transfer.sol]: Let's combine these two conditionals?

Comment (Recmo) [reply packages/contracts/src/contracts/current/protocol/AssetProxy/MixinERC20Transfer.sol]: Let's get rid of all of them!
```
            success := and(success, or(
                iszero(returndatasize),
                and(
                    eq(returndatasize, 32),
                    gt(mload(cdStart), 0)
                )
            ))
```
:D

File: packages/contracts/src/contracts/current/protocol/AssetProxy/interfaces/IAssetData.sol (hunk 1 above)
Comment (Recmo) [reply packages/contracts/src/contracts/current/protocol/AssetProxy/interfaces/IAssetData.sol]: I added documentation explaining the purpose.

File: packages/contracts/src/contracts/current/protocol/Exchange/MixinAssetProxyDispatcher.sol [hunk 4]
```diff
@@ -99,20 +100,67 @@ contract MixinAssetProxyDispatcher is
     {
         // Do nothing if no amount should be transferred.
         if (amount > 0) {
+            require(assetData.length >= 4, "ASSET_DATA_LENGTH");
+            
             // Lookup assetProxy
+            bytes4 assetProxyId;
+            assembly {
+                assetProxyId := and(mload(add(assetData, 32)),
+0xFFFFFFFF00000000000000000000000000000000000000000000000000000000
```
Comment (abandeali1) [inline packages/contracts/src/contracts/current/protocol/Exchange/MixinAssetProxyDispatcher.sol]: Formatting is a bit wonky here.

File: packages/contracts/src/contracts/current/protocol/Exchange/MixinAssetProxyDispatcher.sol [hunk 5]
```diff
@@ -99,20 +100,67 @@ contract MixinAssetProxyDispatcher is
...
+            }
             IAssetProxy assetProxy = assetProxies[assetProxyId];
             // Ensure that assetProxy exists
             require(
                 assetProxy != address(0),
                 ASSET_PROXY_DOES_NOT_EXIST
             );
-            // transferFrom will either succeed or throw.
+            
+            /*
```
Comment (abandeali1) [inline packages/contracts/src/contracts/current/protocol/Exchange/MixinAssetProxyDispatcher.sol]: I think this comment is supposed to be deleted.

File: packages/contracts/src/contracts/current/utils/LibBytes/LibBytes.sol [hunk 6]
```diff
@@ -22,14 +22,14 @@ library LibBytes {
...
     // Revert reasons
-    string constant GREATER_THAN_ZERO_LENGTH_REQUIRED = "GREATER_THAN_ZERO_LENGTH_REQUIRED";
-    string constant GREATER_OR_EQUAL_TO_4_LENGTH_REQUIRED = "GREATER_OR_EQUAL_TO_4_LENGTH_REQUIRED";
-    string constant GREATER_OR_EQUAL_TO_20_LENGTH_REQUIRED = "GREATER_OR_EQUAL_TO_20_LENGTH_REQUIRED";
-    string constant GREATER_OR_EQUAL_TO_32_LENGTH_REQUIRED = "GREATER_OR_EQUAL_TO_32_LENGTH_REQUIRED";
-    string constant GREATER_OR_EQUAL_TO_NESTED_BYTES_LENGTH_REQUIRED = "GREATER_OR_EQUAL_TO_NESTED_BYTES_LENGTH_REQUIRED";
-    string constant GREATER_OR_EQUAL_TO_SOURCE_BYTES_LENGTH_REQUIRED = "GREATER_OR_EQUAL_TO_SOURCE_BYTES_LENGTH_REQUIRED";
-    string constant FROM_LESS_THAN_TO_REQUIRED = "FROM_LESS_THAN_TO_REQUIRED";
-    string constant TO_LESS_THAN_LENGTH_REQUIRED = "TO_LESS_THAN_LENGTH_REQUIRED";
+    // string constant GREATER_THAN_ZERO_LENGTH_REQUIRED = "GREATER_THAN_ZERO_LENGTH_REQUIRED";
```
Comment (hysz) [inline packages/contracts/src/contracts/current/utils/LibBytes/LibBytes.sol]: Should probably add a note here

File: packages/contracts/test/asset_proxy/proxies.ts [hunk 7]
```diff
@@ -244,9 +234,9 @@ describe('Asset Transfer Proxies', () => {
             });
         });
 
-        it('should have an id of 1', async () => {
+        it('should have an id of 0xf47261b0', async () => {
```
Comment (hysz) [inline packages/contracts/test/asset_proxy/proxies.ts]: We should probably make this a constant -- could use the value from `AssetProxyId` enum.

File: packages/contracts/test/asset_proxy/proxies.ts [hunk 8]
```diff
@@ -506,9 +482,9 @@ describe('Asset Transfer Proxies', () => {
             });
         });
 
-        it('should have an id of 2', async () => {
+        it('should have an id of 0x08e937fa', async () => {
```
Comment (hysz) [inline packages/contracts/test/asset_proxy/proxies.ts]: Constant here as well -- could use the value from `AssetProxyId` enum.

**Pull request merged**